src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -614,23 +618,105 @@ public void WithProjectDefaultNamespace()
+
+            SolutionTestHelpers.TestProperty(
+                solution,
+                (s, value) => s.WithProjectChecksumAlgorithm(projectId, value),
+                s => s.GetRequiredProject(projectId).State.ChecksumAlgorithm,
+                SourceHashAlgorithms.Default,
+                defaultThrows: false);
+        }
+
+        [Fact]
+        public void WithProjectChecksumAlgorithm_DocumentUpdates()
+        {
+            var projectId = ProjectId.CreateNewId();
+            var documentAId = DocumentId.CreateNewId(projectId);
+            var documentBId = DocumentId.CreateNewId(projectId);
+            var documentCId = DocumentId.CreateNewId(projectId);
+            var documentDId = DocumentId.CreateNewId(projectId);
+
+            var fileD = Temp.CreateFile();
+            var bytes = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true).GetBytes("Text");
+            fileD.WriteAllBytes(bytes);
+
+            var sha256 = SHA256.Create();
+            var sha1 = SHA1.Create();
+            var checksumSHA1 = sha1.ComputeHash(bytes);
+            var checksumSHA256 = sha256.ComputeHash(bytes);
+            sha256.Dispose();
+            sha1.Dispose();
+
+            using var workspace = CreateWorkspace();
+
+            var textLoaderA = new TestTextLoader("class A {}", SourceHashAlgorithm.Sha1);
+            var textC = SourceText.From("class C {}", encoding: null, checksumAlgorithm: SourceHashAlgorithm.Sha1);
+
+            var solution = workspace.CurrentSolution.
+                AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp).
+                WithProjectChecksumAlgorithm(projectId, SourceHashAlgorithm.Sha1);
+
+            solution = solution.AddDocument(documentAId, "a.cs", textLoaderA);
+            solution = solution.AddDocument(documentBId, "b.cs", "class B {}");
+            solution = solution.AddDocument(documentCId, "c.cs", textC);
+            solution = solution.AddDocument(documentDId, "d.cs", new FileTextLoader(fileD.Path, defaultEncoding: null, SourceHashAlgorithm.Sha1));
+
+            Assert.Equal(SourceHashAlgorithm.Sha1, solution.GetRequiredDocument(documentAId).GetTextSynchronously(default).ChecksumAlgorithm);
+            Assert.Equal(SourceHashAlgorithm.Sha1, solution.GetRequiredDocument(documentBId).GetTextSynchronously(default).ChecksumAlgorithm);
+            Assert.Equal(SourceHashAlgorithm.Sha1, solution.GetRequiredDocument(documentCId).GetTextSynchronously(default).ChecksumAlgorithm);
+
+            var fileText = solution.GetRequiredDocument(documentDId).GetTextSynchronously(default);
+            Assert.Equal(SourceHashAlgorithm.Sha1, fileText.ChecksumAlgorithm);
+            Assert.Equal(checksumSHA1, fileText.GetChecksum());
+
+            solution = solution.WithProjectChecksumAlgorithm(projectId, SourceHashAlgorithm.Sha256);
+
+            // only file loader based documents support updating checksum alg:
```

Comment:
I guess why wouldn't we have the other ones get updated too? I would have thought that'd help in some cases for E&C since it at least meant there was a chance checksums could line up in other cases?

Reply:
Because they can't update the actual checksum.